Leadership transfer cmd

dhiaayachi opened the pull request:
This is an attempt to implement a command that would trigger a raft leadership transfer. 
The idea about this trace back to a customer issue where we needed to trigger leader transfer to be able to upgrade a leader node. 

This dependent to a change in raft that fix some of the gaps in leadership transfer, those gaps would make the transfer fail. https://github.com/hashicorp/raft/pull/487

I made an attempt in naming the sub command but open to any suggestions.

Commit:
add leadership transfer command

Commit:
add RPC call test (flaky)

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=12182)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=12182)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**3** out of **5** committers have signed the CLA.
- [x] andrewstucki
- [x] claire-labry
- [x] boruszak
- [ ] Dhia Ayachi
- [ ] FFMMM

----

**Dhia Ayachi** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA.
If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).


<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=12182).</sub>

andrewstucki committed:
Add doc for proposed annotation (#12716)

claire-labry committed:
change action to pull v1 instead of main

boruszak committed:
Update to admin-partitions.mdx

Removed references to modules in callout box at the top of the page. 

Edit requested by Neena Pemmaraju via Slack on 4/21/22

claire-labry committed:
add changelog entry

boruszak committed:
Update to audit-logging.mdx

Removed references to modules in callout box at the top of the page.

Edit requested by Neena Pemmaraju via Slack on 4/21/22

boruszak committed:
Update to federation.mdx

Removed references to modules in callout box at the top of the page.

Edit requested by Neena Pemmaraju via Slack on 4/21/22

boruszak committed:
Update to namespaces.mdx

Removed references to modules in callout box at the top of the page.

Edit requested by Neena Pemmaraju via Slack on 4/21/22

boruszak committed:
Update to network-segments.mdx

Removed references to modules in callout box at the top of the page.

Edit requested by Neena Pemmaraju via Slack on 4/21/22

boruszak committed:
Update to read-scale.mdx

Removed references to modules in callout box at the top of the page.

Edit requested by Neena Pemmaraju via Slack on 4/21/22

boruszak committed:
Update to redundancy.mdx

Removed references to modules in callout box at the top of the page.

Edit requested by Neena Pemmaraju via Slack on 4/21/22

boruszak committed:
Update to sentinel.mdx

Removed references to modules in callout box at the top of the page.

Edit requested by Neena Pemmaraju via Slack on 4/21/22

FFMMM committed:
add a general pull request template (#12827)

Co-authored-by: Dan Upton <daniel@floppy.co>

Commit:
update raft to v1.3.8 (#12844)

* update raft to v1.3.7

* add changelog

* fix compilation error

* fix HeartbeatTimeout

* fix ElectionTimeout to reload only if value is valid

* fix default values for `ElectionTimeout` and `HeartbeatTimeout`

* fix test defaults

* bump raft to v1.3.8

Commit:
add missing import

Commit:
Merge branch 'main' into leadership-transfer-cmd

Commit:
add changelog

hc-github-team-consul-core committed:
Merge branch 'main' into leadership-transfer-cmd

hc-github-team-consul-core committed:
add command registration

Commit:
Apply suggestions from code review

Co-authored-by: Matt Keeler <mkeeler@users.noreply.github.com>

Commit:
add the possibility of providing an id to raft leadership transfer. Add few tests.

Commit:
Merge branch 'main' into leadership-transfer-cmd

dhiaayachi commented:
Closed in favour of https://github.com/hashicorp/consul/pull/14132
